src/lib/krad/packet.c

```diff
@@ -222,6 +224,162 @@ packet_set_attrset(krb5_context ctx, const char *secret, krad_packet *pkt)
     return kr_attrset_decode(ctx, &tmp, secret, pkt_auth(pkt), &pkt->attrset);
 }
 
+/* Determines if a request or response requires a Message-Authenticator
+ * attribute. */
+static inline krb5_boolean
+requires_msgauth(const char *secret, krad_code code)
+{
+    /* If no secret is provided, assume that the transport is a UNIX socket.
+     * Message-Authenticator is required only on UDP and TCP connections. */
+    if (*secret == '\0')
+        return FALSE;
+
+    /*
+     * draft-ietf-radext-deprecating-radius-03 (section 7.2):
+     *   RADIUS clients MUST include the Message-Authenticator in all
+     *   Access-Request packets when UDP or TCP transport is used.
+     *
+     * draft-ietf-radext-deprecating-radius-03 (section 5.2.4):
+     *   Servers MUST add Message-Authenticator as the first attribute in all
+     *   responses to Access-Request packets.  That is, all Access-Accept,
+     *   Access-Reject, Access-Challenge, and Protocol-Error packets.
+     */
+    return code == krad_code_name2num("Access-Request") ||
+        code == krad_code_name2num("Access-Reject") ||
+        code == krad_code_name2num("Access-Accept") ||
+        code == krad_code_name2num("Access-Challenge") ||
+        code == krad_code_name2num("Protocol-Error");
+}
+
+/* Check if the packet has a Message-Authenticator attribute. */
+static inline krb5_boolean
+has_pkt_msgauth(const krad_packet *pkt)
+{
+    krad_attr msgauth_type = krad_attr_name2num("Message-Authenticator");
+
+    return krad_attrset_get(pkt->attrset, msgauth_type, 0) != NULL;
+}
+
+/* Return the beginning of the Message-Authenticator attribute in pkt, or NULL
+ * if no such attribute is present. */
+static const uint8_t *
+lookup_msgauth_addr(const krad_packet *pkt)
+{
+    krad_attr msgauth_type = krad_attr_name2num("Message-Authenticator");
+    size_t i;
+    uint8_t *p;
+
+    i = OFFSET_ATTR;
+    while (i + 2 < pkt->pkt.length) {
+        p = (uint8_t *)offset(&pkt->pkt, i);
+        if (msgauth_type == (krad_attr)*p)
+            return p;
+        i += p[1];
+    }
+
+    return NULL;
+}
+
+/*
+ * Calculate a message authenticator for pkt as specified in RFC 2869 section
+ * 5.14, placing the result in msgauth_out.  If auth is non-null, use it
+ * instead of the authenticator in pkt, and assume that the first attribute in
+ * pkt is Message-Authenticator.
+ */
+static krb5_error_code
+calculate_msgauth(const char *secret, const krad_packet *pkt,
+                  uint8_t auth[AUTH_FIELD_SIZE],
+                  uint8_t msgauth_out[MD5_DIGEST_SIZE])
```

Comment:
I think that `signature` was actually a better name here because we are not generating the full Message-Authenticator (code + length + signature/keyed checksum), it's only the signature part.

It's actually how it's called [in the RFC](https://datatracker.ietf.org/doc/html/rfc2869#page-34):
> When the checksum is calculated the signature string should be
> considered to be sixteen octets of zero.